File: reference/typescript/src/boolean.ts
import type { Manifold, ManifoldToplevel } from 'manifold-3d';
import { SSKError } from './error.js';
import type { MeshData, ResolvedPiece } from './types.js';

let manifoldModule: Promise<ManifoldToplevel> | null = null;
type Operation = 'union' | 'intersection' | 'difference';
type BooleanContext = { description: string; left: string; right: string };

export async function evaluate(pieces: ResolvedPiece[], meshes: Map<number, MeshData | null>, options: { strict?: boolean } = {}): Promise<MeshData | null> {
  const strict = options.strict ?? true;
  const module = await getManifoldModule();
  const sorted = [...pieces].sort((a, b) => a.id - b.id);
  const active = new Map<number, Manifold>();

  for (const piece of sorted) {
    const mesh = meshes.get(piece.id);
    if (mesh && mesh.faces.length > 0) active.set(piece.id, toManifold(module, mesh, piece));
  }
  if (!active.size) return null;

  const modeOf = new Map<number, string>();
  const addIds: number[] = [];
  for (const piece of sorted) {
    if (!active.has(piece.id)) continue;
    const mode = piece.mode ?? 'add';
    modeOf.set(piece.id, mode);
    if (mode === 'add') addIds.push(piece.id);
  }

  const contrib = new Map<number, Manifold>();
  for (const id of addIds) contrib.set(id, active.get(id)!);

  for (const piece of sorted) {
    if (!active.has(piece.id) || modeOf.get(piece.id) !== 'intersect') continue;
    const affects = piece.affects ? new Set(piece.affects) : null;
    const targets = sorted
      .filter((other) => other.id !== piece.id && active.has(other.id) && ['add', 'intersect'].includes(modeOf.get(other.id) ?? 'add') && (!affects || affects.has(other.id)))
      .map((other) => [other.id, active.get(other.id)!] as const);
    if (!targets.length) continue;
    let union = targets[0][1];
    for (const [targetId, target] of targets.slice(1)) {
      union = op(
        union,
        target,
        'union',
        strict,
        {
          description: `intersect piece ${piece.id} candidate union with piece ${targetId}`,
          left: `intersect piece ${piece.id} candidate union`,
          right: `piece ${targetId}`,
        },
      );
    }
    const result = op(
      active.get(piece.id)!,
      union,
      'intersection',
      strict,
      {
        description: `intersect piece ${piece.id} with candidate union`,
        left: `piece ${piece.id}`,
        right: `intersect piece ${piece.id} candidate union`,
      },
    );
    if (meshFromManifold(result).faces.length > 0) contrib.set(piece.id, result);
  }

  if (!contrib.size) return null;

  for (const piece of sorted) {
    if (!active.has(piece.id) || modeOf.get(piece.id) !== 'subtract') continue;
    const affects = piece.affects ? new Set(piece.affects) : null;
    for (const id of [...contrib.keys()]) {
      if (affects && !affects.has(id)) continue;
      const result = op(
        contrib.get(id)!,
        active.get(piece.id)!,
        'difference',
        strict,
        {
          description: `subtract piece ${piece.id} from piece ${id}`,
          left: `piece ${id} result`,
          right: `subtract piece ${piece.id}`,
        },
      );
      const mesh = meshFromManifold(result);
      if (mesh.faces.length > 0) contrib.set(id, result);
      else contrib.delete(id);
    }
  }

  if (!contrib.size) return null;
  const ids = [...contrib.keys()].sort((a, b) => a - b);
  let result = contrib.get(ids[0])!;
  for (const id of ids.slice(1)) {
    result = op(
      result,
      contrib.get(id)!,
      'union',
      strict,
      {
        description: `final union with piece ${id}`,
        left: 'current final result',
        right: `piece ${id} result`,
      },
    );
  }
  return meshFromManifold(result);
}

async function getManifoldModule(): Promise<ManifoldToplevel> {
  manifoldModule ??= import('manifold-3d')
    .then(({ default: initManifold }) => initManifold())
    .then((module) => {
      module.setup();
      return module;
    });
  return manifoldModule;
}

function toManifold(module: ManifoldToplevel, mesh: MeshData, piece: ResolvedPiece): Manifold {
  const vertProperties = new Float32Array(mesh.vertices.length * 3);
  mesh.vertices.forEach((vertex, index) => vertProperties.set(vertex, index * 3));
  const triVerts = new Uint32Array(mesh.faces.length * 3);
  mesh.faces.forEach((face, index) => triVerts.set(face, index * 3));
  try {
    const manifold = module.Manifold.ofMesh(new module.Mesh({ numProp: 3, vertProperties, triVerts }));
    assertManifoldOk(manifold);
    return manifold;
  } catch (error) {
    throw new SSKError(
      `piece ${piece.id} (${piece.mode ?? 'add'}): manifold import failed: ${errorText(error)}; ${meshDiagnostics(mesh)}`,
    );
  }
}

function meshFromManifold(manifold: Manifold): MeshData {
  const mesh = manifold.getMesh();
  const vertices: number[][] = [];
  for (let i = 0; i < mesh.vertProperties.length; i += mesh.numProp) {
    vertices.push([mesh.vertProperties[i], mesh.vertProperties[i + 1], mesh.vertProperties[i + 2]]);
  }
  const faces: number[][] = [];
  for (let i = 0; i < mesh.triVerts.length; i += 3) {
    faces.push([mesh.triVerts[i], mesh.triVerts[i + 1], mesh.triVerts[i + 2]]);
  }
  return { vertices, faces };
}

function op(a: Manifold, b: Manifold, operation: Operation, strict: boolean, context: BooleanContext): Manifold {
  try {
    const result = operation === 'union' ? a.add(b) : operation === 'intersection' ? a.intersect(b) : a.subtract(b);
    assertManifoldOk(result);
    return result;
  } catch (error) {
    if (strict) {
      throw new SSKError(
        `boolean ${operation} failed: ${context.description}; ${errorText(error)}; ` +
        `${manifoldDiagnostics(context.left, a)}; ${manifoldDiagnostics(context.right, b)}`,
      );
    }
    return operation === 'union' ? a.add(b) : a;
  }
}

function assertManifoldOk(manifold: Manifold): void {
  const status = manifold.status();
  if (status !== 'NoError') throw new SSKError(`manifold status ${status}`);
}

function errorText(error: unknown): string {
  if (error instanceof Error) {
    const withCode = error as Error & { code?: unknown };
    const code = typeof withCode.code === 'string' ? ` (${withCode.code})` : '';
    return `${error.message || error.name}${code}`.replace(/\s+/g, ' ').trim();
  }
  return String(error).replace(/\s+/g, ' ').trim();
}

function meshDiagnostics(mesh: MeshData): string {
  const diagnostics = rawMeshDiagnostics(mesh);
  return [
    `mesh: ${mesh.vertices.length} vertices, ${mesh.faces.length} triangles`,
    `non_finite_vertices=${diagnostics.nonFiniteVertices}`,
    `invalid_faces=${diagnostics.invalidFaces}`,
    `degenerate_faces=${diagnostics.degenerateFaces}`,
    `boundary_edges=${diagnostics.boundaryEdges}`,
    `non_manifold_edges=${diagnostics.nonManifoldEdges}`,
    `bounds=${boundsText(mesh.vertices)}`,
  ].join(', ');
}

function manifoldDiagnostics(label: string, manifold: Manifold): string {
  const parts = [
    `${label}: ${manifold.numVert()} vertices`,
    `${manifold.numTri()} triangles`,
    `status=${manifold.status()}`,
  ];
  try {
    const bounds = manifold.boundingBox();
    parts.push(`bounds=min=${vecText(bounds.min)}, max=${vecText(bounds.max)}`);
  } catch {
    parts.push('bounds=unavailable');
  }
  return parts.join(', ');
}

function rawMeshDiagnostics(mesh: MeshData): {
  nonFiniteVertices: number;
  invalidFaces: number;
  degenerateFaces: number;
  boundaryEdges: number;
  nonManifoldEdges: number;
} {
  let nonFiniteVertices = 0;
  let invalidFaces = 0;
  let degenerateFaces = 0;
  const edgeCounts = new Map<string, number>();

  mesh.vertices.forEach((vertex) => {
    if (!Number.isFinite(vertex[0]) || !Number.isFinite(vertex[1]) || !Number.isFinite(vertex[2])) nonFiniteVertices += 1;
  });

  for (const face of mesh.faces) {
    if (face.length !== 3 || face.some((index) => !Number.isInteger(index) || index < 0 || index >= mesh.vertices.length)) {
      invalidFaces += 1;
      continue;
    }
    if (new Set(face).size < 3 || triangleAreaSquared(mesh.vertices, face) <= 1e-24) degenerateFaces += 1;
    for (const [a, b] of [[face[0], face[1]], [face[1], face[2]], [face[2], face[0]]] as const) {
      const key = a < b ? `${a},${b}` : `${b},${a}`;
      edgeCounts.set(key, (edgeCounts.get(key) ?? 0) + 1);
    }
  }

  let boundaryEdges = 0;
  let nonManifoldEdges = 0;
  for (const count of edgeCounts.values()) {
    if (count === 1) boundaryEdges += 1;
    else if (count > 2) nonManifoldEdges += 1;
  }

  return { nonFiniteVertices, invalidFaces, degenerateFaces, boundaryEdges, nonManifoldEdges };
}

function triangleAreaSquared(vertices: number[][], face: number[]): number {
  const a = vertices[face[0]];
  const b = vertices[face[1]];
  const c = vertices[face[2]];
  const ab = [b[0] - a[0], b[1] - a[1], b[2] - a[2]];
  const ac = [c[0] - a[0], c[1] - a[1], c[2] - a[2]];
  const cross = [
    ab[1] * ac[2] - ab[2] * ac[1],
    ab[2] * ac[0] - ab[0] * ac[2],
    ab[0] * ac[1] - ab[1] * ac[0],
  ];
  return cross[0] ** 2 + cross[1] ** 2 + cross[2] ** 2;
}

function boundsText(vertices: number[][]): string {
  const finite = vertices.filter((vertex) => Number.isFinite(vertex[0]) && Number.isFinite(vertex[1]) && Number.isFinite(vertex[2]));
  if (!finite.length) return 'unavailable';
  const min = [...finite[0]];
  const max = [...finite[0]];
  for (const vertex of finite.slice(1)) {
    for (let axis = 0; axis < 3; axis += 1) {
      min[axis] = Math.min(min[axis], vertex[axis]);
      max[axis] = Math.max(max[axis], vertex[axis]);
    }
  }
  return `min=${vecText(min)}, max=${vecText(max)}`;
}

function vecText(value: number[] | { x: number; y: number; z: number }): string {
  const vector = Array.isArray(value) ? value : [value.x, value.y, value.z];
  return `(${vector.map(formatNumber).join(', ')})`;
}

function formatNumber(value: number): string {
  return Number.isFinite(value) ? Number(value.toPrecision(6)).toString() : String(value);
}

File: reference/python/ssklib/boolean.py
"""CSG boolean evaluation."""

from collections import Counter
from typing import NamedTuple

import trimesh

from .error import SSKError


class _OpContext(NamedTuple):
    description: str
    left: str
    right: str


def evaluate(pieces: list, meshes: dict, *, strict: bool = True) -> trimesh.Trimesh:

    pieces = sorted(pieces, key=lambda piece: piece['id'])

    active = {}
    for p in pieces:
        pid = p['id']
        if pid in meshes and meshes[pid] is not None:
            v, f = meshes[pid]
            active[pid] = trimesh.Trimesh(vertices=v, faces=f, process=False)
    if not active:
        return trimesh.Trimesh()

    mode_of = {}
    add_ids = []
    for p in pieces:
        pid = p['id']
        if pid not in active:
            continue
        m = p.get('mode', 'add')
        mode_of[pid] = m
        if m == 'add':
            add_ids.append(pid)

    contrib = {pid: active[pid] for pid in add_ids}

    for p in pieces:
        pid = p['id']
        if pid not in active or mode_of.get(pid) != 'intersect':
            continue
        aff = p.get('affects')
        targets = [(o['id'], active[o['id']]) for o in pieces
                   if o['id'] != pid and o['id'] in active
                   and mode_of.get(o['id']) in ('add', 'intersect')
                   and (aff is None or o['id'] in aff)]
        if not targets:
            continue
        union = targets[0][1]
        for target_id, t in targets[1:]:
            union = _op(
                union,
                t,
                'union',
                strict=strict,
                context=_OpContext(
                    f"intersect piece {pid} candidate union with piece {target_id}",
                    f"intersect piece {pid} candidate union",
                    f"piece {target_id}",
                ),
            )
            if union is None:
                break
        if union is not None:
            res = _op(
                active[pid],
                union,
                'intersection',
                strict=strict,
                context=_OpContext(
                    f"intersect piece {pid} with candidate union",
                    f"piece {pid}",
                    f"intersect piece {pid} candidate union",
                ),
            )
            if res is not None and len(res.faces) > 0:
                contrib[pid] = res

    if not contrib:
        return trimesh.Trimesh()

    for p in pieces:
        pid = p['id']
        if pid not in active or mode_of.get(pid) != 'subtract':
            continue
        aff = p.get('affects')
        for cid in list(contrib):
            if aff is not None and cid not in aff:
                continue
            res = _op(
                contrib[cid],
                active[pid],
                'difference',
                strict=strict,
                context=_OpContext(
                    f"subtract piece {pid} from piece {cid}",
                    f"piece {cid} result",
                    f"subtract piece {pid}",
                ),
            )
            if res is not None:
                if len(res.faces) > 0:
                    contrib[cid] = res
                else:
                    del contrib[cid]

    if not contrib:
        return trimesh.Trimesh()

    # union remaining
    ids = sorted(contrib)
    result = contrib[ids[0]]
    for cid in ids[1:]:
        r = _op(
            result,
            contrib[cid],
            'union',
            strict=strict,
            context=_OpContext(
                f"final union with piece {cid}",
                "current final result",
                f"piece {cid} result",
            ),
        )
        if r is not None:
            result = r
    return result


def _op(
    a: trimesh.Trimesh,
    b: trimesh.Trimesh,
    op: str,
    *,
    strict: bool,
    context: _OpContext,
):
    fn = {'union': trimesh.boolean.union,
          'intersection': trimesh.boolean.intersection,
          'difference': trimesh.boolean.difference}[op]
    try:
        return fn([a, b], engine='manifold')
    except Exception as manifold_error:
        try:
            return fn([a, b])
        except Exception as second_error:
            if strict:
                raise SSKError(_failure_message(
                    op,
                    context,
                    manifold_error,
                    second_error,
                    _mesh_report(context.left, a),
                    _mesh_report(context.right, b),
                )) from second_error
            return trimesh.util.concatenate([a, b]) if op == 'union' else a


def _failure_message(op, context, manifold_error, fallback_error, left_report, right_report):
    return (
        f"boolean {op} failed: {context.description}; "
        f"manifold={_error_text(manifold_error)}; "
        f"fallback={_error_text(fallback_error)}; "
        f"{left_report}; {right_report}"
    )


def _error_text(error: Exception) -> str:
    message = str(error).strip()
    if not message:
        message = error.__class__.__name__
    return message.replace('\n', ' ')


def _mesh_report(label: str, mesh: trimesh.Trimesh) -> str:
    boundary_edges, non_manifold_edges, degenerate_faces = _edge_diagnostics(mesh)
    parts = [
        f"{label}: {len(mesh.vertices)} vertices, {len(mesh.faces)} triangles",
        f"watertight={mesh.is_watertight}",
        f"winding_consistent={mesh.is_winding_consistent}",
        f"volume={mesh.is_volume}",
        f"degenerate_faces={degenerate_faces}",
        f"boundary_edges={boundary_edges}",
        f"non_manifold_edges={non_manifold_edges}",
    ]
    bounds = _bounds_text(mesh)
    if bounds is not None:
        parts.append(f"bounds={bounds}")
    return ', '.join(parts)


def _edge_diagnostics(mesh: trimesh.Trimesh) -> tuple[int, int, int]:
    edge_counts = Counter()
    degenerate_faces = 0
    vertex_count = len(mesh.vertices)
    for face in mesh.faces:
        face_ids = [int(index) for index in face]
        if len(set(face_ids)) < 3:
            degenerate_faces += 1
            continue
        if any(index < 0 or index >= vertex_count for index in face_ids):
            continue
        for start, end in ((0, 1), (1, 2), (2, 0)):
            a, b = face_ids[start], face_ids[end]
            edge_counts[tuple(sorted((a, b)))] += 1
    boundary_edges = sum(1 for count in edge_counts.values() if count == 1)
    non_manifold_edges = sum(1 for count in edge_counts.values() if count > 2)
    return boundary_edges, non_manifold_edges, degenerate_faces


def _bounds_text(mesh: trimesh.Trimesh) -> str | None:
    try:
        bounds = mesh.bounds
    except Exception:
        return None
    if bounds is None or len(bounds) != 2:
        return None
    return f"min={_vec_text(bounds[0])}, max={_vec_text(bounds[1])}"


def _vec_text(values) -> str:
    return '(' + ', '.join(f"{float(value):.6g}" for value in values) + ')'
